File: src/entities/contract-definition.ts
import { Criterion, CriterionInput } from "./criterion";
import { JsonLdId } from "./jsonld";

export class ContractDefinition extends JsonLdId {
  get accessPolicyId(): string {
    return this.mandatoryValue('edc', 'accessPolicyId');
  }

  get contractPolicyId(): string {
    return this.mandatoryValue('edc', 'contractPolicyId');
  }

  get assetsSelector(): Criterion[] {
    return this.arrayOf(() => new Criterion(), 'edc', 'assetsSelector')
  }
}

export interface ContractDefinitionInput {
  "@id"?: string,
  accessPolicyId: string;
  contractPolicyId: string;
  assetsSelector: CriterionInput[];
}

File: rust/src/entities/contract_definition.rs
use serde::{Deserialize, Serialize};

use super::criterion::Criterion;

#[derive(Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ContractDefinition {
    pub id: String,
    pub access_policy_id: String,
    pub contract_policy_id: String,
    pub criteria: Vec<Criterion>,
}

pub type ContractDefinitionInput = ContractDefinition;
